api/org/commcare/api/persistence/SqlIndexedStorageUtility.java

```diff
@@ -0,0 +1,388 @@
+            e.printStackTrace();
+        }
+    }
+
+    public T readFromBytes(byte[] mBytes) {
+
+        T returnPrototype;
+        ByteArrayInputStream mByteStream = null;
+        try {
+            returnPrototype = prototype.newInstance();
+            mByteStream = new ByteArrayInputStream(mBytes);
+            returnPrototype.readExternal(new DataInputStream(mByteStream), PrototypeManager.getDefault());
+            return returnPrototype;
+        } catch (InstantiationException | IllegalAccessException | DeserializationException | IOException e) {
+            e.printStackTrace();
+        } finally{
+            if(mByteStream != null){
+                try {
+                    mByteStream.close();
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public T read(int id) {
+        byte[] mBytes = readBytes(id);
+        return readFromBytes(mBytes);
+
+    }
+
+    @Override
+    public Vector getIDsForValue(String fieldName, Object value) {
+        Connection c = null;
+        PreparedStatement preparedStatement = null;
+        try {
+            c = this.getConnection();
+            preparedStatement = SqlHelper.prepareTableSelectStatement(c, this.tableName,
+                    new String[]{fieldName}, new String[]{(String)value}, prototype.newInstance());
+            if (preparedStatement == null) {
+                return null;
+            }
+            ResultSet rs = preparedStatement.executeQuery();
+                    Vector<Integer> ids = new Vector<>();
+            while (rs.next()) {
+                ids.add(rs.getInt(org.commcare.modern.database.DatabaseHelper.ID_COL));
+            }
+            return ids;
+        } catch (InstantiationException | IllegalAccessException | SQLException | ClassNotFoundException e) {
+            e.printStackTrace();
+        } finally {
+            try {
+                if (c != null) {
+                    c.close();
+                }
+                if (preparedStatement != null) {
+                    preparedStatement.close();
+                }
+            } catch (SQLException e) {
+                e.printStackTrace();
+            }
+
+        }
+
+        return null;
+    }
+
+    @Override
+    public T getRecordForValue(String fieldName, Object value)
+            throws NoSuchElementException, InvalidIndexException {
+        Connection c = null;
+        PreparedStatement preparedStatement = null;
+        try {
+            c = this.getConnection();
+            preparedStatement =
+                    SqlHelper.prepareTableSelectStatement(c, this.tableName,
+                            new String[]{fieldName}, new String[]{(String)value},
+                            prototype.newInstance());
+            if (preparedStatement == null) {
+                throw new NoSuchElementException();
```

Comment:
That doesn't sound correct to me. Why would the preparedStatement being null imply that there was no element? Doesn't the preparedstatement being null mean that there was an error with the syntax?